substrate/primitives/virtualization/src/native.rs

```diff
@@ -0,0 +1,331 @@
+		let result = match &*rc.borrow() {
+			Memory::Idle(instance) => instance.read_memory_into_slice(offset, dest),
+			Memory::Executing(caller) => caller.read_memory_into_slice(offset, dest),
+		};
+		result.map(|_| ()).map_err(|_| MemoryError::OutOfBounds)
+	}
+
+	fn write(&mut self, offset: u32, src: &[u8]) -> Result<(), MemoryError> {
+		let rc = self.upgrade().ok_or(MemoryError::InvalidInstance)?;
+		let result = match &mut *rc.borrow_mut() {
+			Memory::Idle(instance) => instance.write_memory(offset, src),
+			Memory::Executing(caller) => caller.write_memory(offset, src),
+		};
+		result.map_err(|_| MemoryError::OutOfBounds)
+	}
+}
+
+impl Virt {
+	/// Return a mutable reference to the state shared with the syscall handler.
+	///
+	/// # SAFETY
+	///
+	/// The caller must make sure that no other reference to [`Self::state`] exists
+	/// while holding the reference returned from this function.
+	///
+	/// # Traps
+	///
+	/// Traps if being called outside of `on_ecall`.
+	unsafe fn state(&mut self) -> &mut SharedState<()> {
+		// # SAFETY
+		//
+		// ## Life Times
+		//
+		// The reference is created from a raw pointer which was in turn created from a
+		// mutable reference passed into [`Self::`execute`]. This makes sure that no other
+		// reference exists while inside `execute`. The pointer is stored within
+		// [`Self::while_exec`] which is only set while being within `execute`.
+		//
+		// ## Change of generic parameter
+		//
+		// We transmute `&mut SharedState<T>` to `&mut SharedState<()>` here. This is safe because
+		// `SharedState` is using #[repr(C)] alignment where the change of the last field will
+		// not impact the alignment of the rest of the fields. Additionally, by chosing a ZST
+		// for `T` we prevent any code that accesses this data from being generated. Hence
+		// no assumptions over `T` will be made.
+		&mut *(self
+			.while_exec
+			.as_mut()
+			.expect(
+				"Is set while executing. This function is only called from on_ecall;
+				on_ecall is only called while executing; qed",
+			)
+			.state as *mut _)
+	}
+
+	fn internal_execute<T>(
+		&mut self,
+		function: &str,
+		syscall_handler: SyscallHandler<T>,
+		state: &mut SharedState<T>,
+	) -> Result<(), ExecError> {
+		let mut state_args = StateArgs::new();
+		state_args.reset_memory(false).set_gas(state.gas_left.try_into().map_err(|_| {
+			log::error!("{} is not a valid gas value", state.gas_left);
+			ExecError::InvalidGasValue
+		})?);
+		self.instance
+			.update_state(state_args)
+			.expect("We only set valid state above; qed");
+
+		// It does not really make sense to set `exit` to true before calling execute. However,
+		// it seems least surprising to not even start the execution in this case.
+		if state.exit {
+			return Ok(())
+		}
+
+		self.while_exec = Some(WhileExec {
+			// SAFETY: `SyscallHandler` and `ErasedSyscallHandler` are ABI compatible
+			syscall_handler: unsafe { mem::transmute(syscall_handler) },
```

Comment:
maybe move the unsafe to a From impl close to the Erased type so we don't have to explain it here.